server/sonar-server/src/main/java/org/sonar/server/user/ws/GroupsAction.java

```diff
@@ -0,0 +1,140 @@
+import org.sonar.api.server.ws.WebService.Param;
+import org.sonar.api.utils.Paging;
+import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.core.user.GroupMembershipDto;
+import org.sonar.core.user.GroupMembershipQuery;
+import org.sonar.core.user.UserDto;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.exceptions.NotFoundException;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+public class GroupsAction implements BaseUsersWsAction {
+
+  private static final String PARAM_LOGIN = "login";
+  private static final String PARAM_SELECTED = "selected";
+  private static final String PARAM_QUERY = "q";
+
+  private static final String SELECTION_ALL = "all";
+  private static final String SELECTION_SELECTED = "selected";
+  private static final String SELECTION_DESELECTED = "deselected";
+
+  private final DbClient dbClient;
+
+  public GroupsAction(DbClient dbClient) {
+    this.dbClient = dbClient;
+  }
+
+  @Override
+  public void define(NewController context) {
+    NewAction action = context.createAction("groups")
+      .setDescription("List the groups a user belongs to.")
+      .setHandler(this)
+      .setResponseExample(getClass().getResource("example-groups.json"))
+      .setSince("5.2");
+
+    action.createParam(PARAM_LOGIN)
+      .setDescription("A user login")
+      .setExampleValue("admin")
+      .setRequired(true);
+
+    action.createParam(PARAM_SELECTED)
+      .setDescription("If specified, only show groups the user is member of (selected) or not (deselected).")
+      .setPossibleValues(SELECTION_SELECTED, SELECTION_DESELECTED, SELECTION_ALL)
+      .setDefaultValue(SELECTION_ALL);
+
+    action.createParam(PARAM_QUERY)
+      .setDescription("If specified, only show groups whose name contains the query.")
+      .setExampleValue("user");
+
+    action.addPagingParams(25);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    String login = request.mandatoryParam(PARAM_LOGIN);
+    UserDto user = dbClient.userDao().selectActiveUserByLogin(login);
+    if (user == null) {
+      throw new NotFoundException(String.format("Unable to find a user with login '%s'", login));
+    }
+
+    int pageSize = request.mandatoryParamAsInt(Param.PAGE_SIZE);
+    int page = request.mandatoryParamAsInt(Param.PAGE);
+    String queryString = request.param(PARAM_QUERY);
+    String selected = request.param(PARAM_SELECTED);
+
+    GroupMembershipQuery query = GroupMembershipQuery.builder()
+      .login(login)
+      .groupSearch(queryString)
+      .membership(getMembership(selected))
+      .pageIndex(page)
+      .pageSize(pageSize)
+      .build();
+
+    int total = dbClient.groupMembershipDao().countGroups(query, user.getId());
+
+    Paging paging = Paging.create(pageSize, page, total);
+
+    List<GroupMembershipDto> groups = dbClient.groupMembershipDao().selectGroups(query, user.getId(), paging.offset(), pageSize);
```

Comment:
You mean org.sonar.server.es.SearchOptions? I don't like the idea of using an "ES" oriented component for RDBMS search, but maybe it can be refactored in some way.